data_processors/pipeline/domain/workflow.py

```diff
@@ -230,12 +234,98 @@ def __init__(self, this_workflow):
+            raise WorkflowRuleError(f"Workflow {self.workflow.type_name} wfr_id: '{self.workflow.wfr_id}' must be "
+                                    f"associated with a SequenceRun. SequenceRun is: {this_sqr}")
         return self
 
     def must_have_output(self):
         # use case e.g. bcl convert workflow run must have output in order to continue next step(s)
         if self.workflow.output is None:
-            raise ValueError(f"Workflow '{self.workflow.wfr_id}' output is None")
+            raise WorkflowRuleError(f"Workflow '{self.workflow.wfr_id}' output is None")
+        return self
+
+
+class MetadataRuleError(ValueError):
+    pass
+
+
+class MetadataRule:
+    """
+    MetadataRule model that check some state must conform in wrapped LabMetadata. Implement your rule that start with
+    must_XX expression. Raise MetadataRuleError if not conformant. Otherwise, return itself for chain validation.
+
+    NOTE: The aspect here is just "metadata" itself as-is validation. For LibraryRun aspect, see LibraryRule.
+    """
+
+    def __init__(self, this_metadata):
+        self.this_metadata = this_metadata
+
+    def must_exist(self):
+        if self.this_metadata is None:
+            raise MetadataRuleError(f"No metadata.")
+        return self
+
+    def must_not_manual(self):
+        from data_portal.models.labmetadata import LabMetadataWorkflow
+        if self.this_metadata.workflow.lower() == LabMetadataWorkflow.MANUAL.value.lower():
+            raise MetadataRuleError(f"Workflow is set to manual.")
+        return self
+
+    def must_not_ntc(self):
+        from data_portal.models.labmetadata import LabMetadataPhenotype
+        if self.this_metadata.phenotype.lower() == LabMetadataPhenotype.N_CONTROL.value.lower():
+            raise MetadataRuleError(f"Negative-control sample.")
+        return self
+
+    def must_be_wgs(self):
+        from data_portal.models.labmetadata import LabMetadataType
+        if self.this_metadata.type.lower() != LabMetadataType.WGS.value.lower():
+            raise MetadataRuleError(f"'WGS' != '{self.this_metadata.type}'.")
+        return self
+
+    def must_be_wts(self):
+        from data_portal.models.labmetadata import LabMetadataType
+        if self.this_metadata.type.lower() != LabMetadataType.WTS.value.lower():
+            raise MetadataRuleError(f"'WTS' != '{self.this_metadata.type}'.")
+        return self
+
+    def must_be_cttso_ctdna(self):
+        from data_portal.models.labmetadata import LabMetadataType, LabMetadataAssay
+        if not (self.this_metadata.type.lower() == LabMetadataType.CT_DNA.value.lower() and
+                self.this_metadata.assay.lower() == LabMetadataAssay.CT_TSO.value.lower()):
+            raise MetadataRuleError(
+                f"Type: 'ctDNA' != '{self.this_metadata.type}' or Assay: 'ctTSO' != '{self.this_metadata.assay}'"
+            )
+        return self
+
+
+class LibraryRuleError(ValueError):
+    pass
+
+
+class LibraryRule:
```

Comment:
That's currently intended for the `LibraryRun` level? In that case I'd rename it?
The checks could apply on library or LibraryRun level. The former would have to be an aggregation of LibraryRun level results, which may be tricky... The latter would build on the results from the QC workflow I suppose.

Reply:
Yep, yep. This `LibraryRule`, at the mo, is still a placeholder. Since, this is still a place-holder, let pend it to later implementor decision on what it might actually do; when starting on this part...

My vision is that; thinking of Library concept as even higher abstraction such that describing it as "Library domain aggregate" comprises of `LabMetadata`, `LibraryRun`, etc. And, very possibly `LibraryRun` as this aggregate root... Guess, I'm corrupted by a bit too much of these. 😁 

- https://martinfowler.com/bliki/DDD_Aggregate.html
- https://stackoverflow.com/questions/1958621/whats-an-aggregate-root

Reply:
HeHe, fair enough!
Perhaps one for the data model refactoring in v2 ;-)

Reply:
Ah well; renamed them all, anyway! Merging... 😁